client/client.go

```diff
@@ -0,0 +1,322 @@
+func DefaultLogger(format string, args ...interface{}) {
+	log.Printf("[DEBUG] "+format, args...)
+}
+
+// RoundTripper satisfies the http.RoundTripper interface and is used to
+// customize the default http client RoundTripper
+type RoundTripper struct {
+	// Default http.RoundTripper
+	Rt http.RoundTripper
+	// Additional request headers to be set (not appended) in all client
+	// requests
+	headers *http.Header
+	// A pointer to a map of headers to be masked in logger
+	maskHeaders *map[string]struct{}
+	// A custom function to format and mask JSON requests and responses
+	FormatJSON func([]byte) (string, error)
+	// How many times HTTP connection should be retried until giving up
+	MaxRetries int
+	// If Logger is not nil, then RoundTrip method will debug the JSON
+	// requests and responses
+	Logger func(string, ...interface{})
+}
+
+// List of headers that contain sensitive data.
+var defaultSensitiveHeaders = map[string]struct{}{
+	"x-auth-token":                    {},
+	"x-auth-key":                      {},
+	"x-service-token":                 {},
+	"x-storage-token":                 {},
+	"x-account-meta-temp-url-key":     {},
+	"x-account-meta-temp-url-key-2":   {},
+	"x-container-meta-temp-url-key":   {},
+	"x-container-meta-temp-url-key-2": {},
+	"set-cookie":                      {},
+	"x-subject-token":                 {},
+}
+
+// GetDefaultSensitiveHeaders returns the default list of headers to be masked
+func GetDefaultSensitiveHeaders() []string {
+	headers := make([]string, len(defaultSensitiveHeaders))
+	i := 0
+	for k := range defaultSensitiveHeaders {
+		headers[i] = k
+		i++
+	}
+
+	return headers
+}
+
+// SetSensitiveHeaders sets the list of case insensitive headers to be masked in
+// debug log
+func (rt *RoundTripper) SetSensitiveHeaders(headers []string) {
+	newHeaders := make(map[string]struct{}, len(headers))
+
+	for _, h := range headers {
+		newHeaders[h] = struct{}{}
+	}
+
+	// this is concurrency safe
+	rt.maskHeaders = &newHeaders
+}
+
+// SetHeaders sets request headers to be set (not appended) in all client
+// requests
+func (rt *RoundTripper) SetHeaders(headers http.Header) {
+	newHeaders := make(http.Header, len(headers))
+	for k, v := range headers {
+		s := make([]string, len(v))
+		for i, v := range v {
+			s[i] = v
+		}
+		newHeaders[k] = s
+	}
+
+	// this is concurrency safe
+	rt.headers = &newHeaders
+}
+
+func (rt *RoundTripper) hideSensitiveHeadersData(headers http.Header) []string {
+	result := make([]string, len(headers))
+	headerIdx := 0
+	// this is concurrency safe
+	maskHeaders := rt.maskHeaders
+	if maskHeaders == nil {
+		maskHeaders = &defaultSensitiveHeaders
+	}
+	for header, data := range headers {
+		if _, ok := (*maskHeaders)[strings.ToLower(header)]; ok {
+			result[headerIdx] = fmt.Sprintf("%s: %s", header, "***")
+		} else {
+			result[headerIdx] = fmt.Sprintf("%s: %s", header, strings.Join(data, " "))
+		}
+		headerIdx++
+	}
+
+	return result
+}
```

Comment:
This might be a little more readable:

```go
func (rt *RoundTripper) hideSensitiveHeadersData(headers http.Header) []string {
  result := make([]string, len(headers))
  headerIdx := 0

  // this is concurrency safe
  v := rt.maskHeaders
  if v == nil {
    v = &defaultSensitiveHeaders
  }
  maskHeaders := *v

  for header, data := range headers {
    v := strings.ToLower(header)
    if _, ok := maskHeaders[v]; ok {
      result[headerIdx] = fmt.Sprintf("%s: %s", header, "***")
    } else {
      result[headerIdx] = fmt.Sprintf("%s: %s", header, strings.Join(data, " "))
    }
    headerIdx++
  }

  return result
}
```